src/nccl_ofi_net.cpp

```diff
@@ -872,7 +872,10 @@ int nccl_net_ofi_domain_t::release_domain(bool skip_device_lock, bool force_clea
 	int ret = 0;
 	nccl_net_ofi_device_t *device_ptr = this->device;
 
-	nccl_net_ofi_mutex_lock(&this->domain_lock);
+	// The caller takes device_lock when force_cleanup.
+	if (!skip_device_lock) {
```

Comment:
I'd really rather we not have this paradigm in the code.  Clang's thread static analysis can't handle conditional locks, so by doing this we're breaking any hope of actually having static analysis here.

Reply:
I agree with you @bwbarrett.  in this PR I've moved the existing conditional lock to the beginning to avoid redundant locks, it's not a new one. Today the conditional locking exists in few places mainly to handle device and domain leaked resource cleanup when the plugin and domain is destroyed. we will have to revisit that logic to avoid the conditional locks but looks like it will take sometime to reimplement cleanly.